### PR TITLE
[Model Averaging] Skip model averaging for the first K steps

### PR DESCRIPTION
**Pull request by wayi1**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#61207**
* #61206

Model averager now must be combined with post-localSGD DDP communication hook. It will skip model averaging for the first K steps, because post-localSGD communication hook will run global gradient averaging during this phase.

Proposal: https://github.com/pytorch/pytorch/issues/59699

Differential Revision: [D29523738](https://our.internmc.facebook.com/intern/diff/D29523738/)

### CODE REVIEW

File: torch/distributed/algorithms/model_averaging/averagers.py [hunk 1]
```diff
@@ -6,7 +6,7 @@
 
 class PeriodicModelAverager:
     r"""
-    Averages parameters periodically or during the warm-up stage.
+    Averages parameters periodically after the warm-up stage.
 
     This can be used for running `post-local SDG <https://arxiv.org/abs/1808.07217>`_,
```
Comment (pritamdamania87) [inline torch/distributed/algorithms/model_averaging/averagers.py:11]: nit: s/SDG/SGD

File: torch/testing/_internal/distributed/distributed_test.py [hunk 2]
```diff
@@ -1007,12 +1007,12 @@ def test_periodic_model_averager(self):
             param = next(model.parameters())
             tensor = torch.ones_like(param.data) * rank
             expected_avg_tensor = torch.ones_like(param.data) * sum(range(world_size)) / world_size
-            averager = averagers.PeriodicModelAverager(model, warmup_steps=10, period=4)
+            averager = averagers.PeriodicModelAverager(model, warmup_steps=12, period=4)
```
Comment (pritamdamania87) [inline torch/testing/_internal/distributed/distributed_test.py:1010]: Shouldn't we have a test that combines PeriodicModelAverager with the local sgd hook?

File: torch/distributed/algorithms/model_averaging/averagers.py [hunk 3]
```diff
@@ -84,10 +88,10 @@ def __init__(
...
     def average_parameters(self):
         r"""
-        Averages parameters if ``step`` is less than ``warmup_steps``,
-        or it can be divided by ``period``, where ``step`` is increased by 1
+        Averages parameters if ``step`` is no less than ``warmup_steps``
+        and it can be divided by ``period``, where ``step`` is increased by 1
         at each iteration in the training loop.
         """
-        if self.step < self.warmup_steps or self.step % self.period == 0:
+        if self.step >= self.warmup_steps and self.step % self.period == 0:
```
Comment (pritamdamania87) [inline torch/distributed/algorithms/model_averaging/averagers.py:95]: If warmup steps is 101 and period is 4, is it correct that we don't average for steps 101 - 103 and then average on step 104? Shouldn't this check be `(self.step - self.warmup_steps) % self.period == 0` ?

Comment (SciPioneer) [reply torch/distributed/algorithms/model_averaging/averagers.py:95]: Fixed and updated the test case to cover this case.

File: torch/testing/_internal/distributed/distributed_test.py (hunk 2 above)
Comment (SciPioneer) [reply torch/testing/_internal/distributed/distributed_test.py:1010]: Hmm, I prefer keeping the unit test self-contained here.

Let me add another integration test called `test_post_localSGD_periodic_averaging` that involves both the comm hook and the model averaging component in the next PR.

**Review by pritamdamania87 — APPROVED** [review]
